vm/src/stdlib/itertools.rs

```diff
@@ -4,14 +4,74 @@ use std::ops::{AddAssign, SubAssign};
 
 use num_bigint::BigInt;
 
-use crate::function::OptionalArg;
+use crate::function::{OptionalArg, PyFuncArgs};
 use crate::obj::objbool;
 use crate::obj::objint::{PyInt, PyIntRef};
 use crate::obj::objiter::{call_next, get_iter, new_stop_iteration};
+use crate::obj::objtype;
 use crate::obj::objtype::PyClassRef;
 use crate::pyobject::{PyClassImpl, PyObjectRef, PyRef, PyResult, PyValue};
 use crate::vm::VirtualMachine;
 
+#[pyclass(name = "chain")]
+#[derive(Debug)]
+struct PyItertoolsChain {
+    iterables: Vec<PyObjectRef>,
+    cur: RefCell<usize>,
+    cur_iter: RefCell<Option<PyObjectRef>>,
```

Comment:
I think you could combine these two `RefCell`s into one holding a tuple (`RefCell<(usize, Option<PyObjectRef>)>`) in order to reduce a bit of the overhead of runtime-checked mutability. Then, in `next()`, you could borrow both of them at the start of the method as so: `let (cur, cur_iter) = self.cur.borrow_mut()`. Then, you won't have any problems with borrows colliding as well.

Reply:
mhm. Tho, I thought it'd better if the code was constructed with as little mutable borrows as possible. (After all, to my understanding, the whole hassle of these `RefCell`s is `PyObject`'s `self` being immutable, which I guess is something we'll change in the future? So reconsutrcting the code with long mutable borrows will just get us more work later.